src/index.ts

```diff
@@ -1,39 +1,57 @@
 import debug = require('debug')
-import flatten = require('array-flatten')
 
 const log = debug('compose-middleware')
 
 export type Next<T = void> = (err?: Error | null) => T
 export type RequestHandler <T, U, V = void> = (req: T, res: U, next: Next<V>) => V
-export type ErrorHandler <T, U, V = void> = (err: Error, req: T, res: U, next: Next<V>) => V
+export type ErrorHandler <T, U, V = void> = (err: Error | null, req: T, res: U, next: Next<V>) => V
 export type Middleware <T, U, V = void> = RequestHandler<T, U, V> | ErrorHandler<T, U, V>
 
-export type Handler <T, U, V = void> = Middleware<T, U, V> | flatten.NestedArray<Middleware<T, U, V>>
+/**
+ * Compose a variadic number of middlewares into a single middleware.
+ */
+export function compose <T, U, V = void> (
+  middleware?: RequestHandler<T, U, V>,
+  ...middlewares: Middleware<T, U, V>[]
+): RequestHandler<T, U, V>
+export function compose <T, U, V = void> (
+  middleware: ErrorHandler<T, U, V>,
+  ...middlewares: Middleware<T, U, V>[]
+): ErrorHandler<T, U, V>
+export function compose <T, U, V = void> (
+  middleware?: Middleware<T, U, V>,
+  ...middlewares: Middleware<T, U, V>[]
+): Middleware <T, U, V> {
+  if (!middleware) {
+    return successes<T, U, V>()
+  } else if (middleware.length === 3) {
+    return successes<T, U, V>(middleware as RequestHandler<T, U, V>, ...middlewares)
+  }
+  return errors<T, U, V>(middleware as ErrorHandler<T, U, V>, ...middlewares)
+}
 
 /**
- * Compose an array of middleware handlers into a single handler.
+ * Compose a variadic number of middlewares into a single success middleware.
  */
-export function compose <T, U, V = void> (...handlers: Handler<T, U, V>[]): RequestHandler<T, U, V> {
-  const middleware = generate(handlers)
+export function successes <T, U, V = void> (...middlewares: Middleware<T, U, V>[]): RequestHandler<T, U, V> {
```

Comment:
Is there a better name for this? Maybe `handler` and `errorHandler`?

Reply:
Are you saying to rename `successes` function to `handler` and `errors` function to `errorHandler`? I'm a little confused about renaming `errors` because a comment above you are talking about backward compatibility. What did you mean?